composer/trainer/trainer.py

```diff
@@ -1325,6 +1353,23 @@ def __init__(
             # Only wrap the module if required
             self.state.model = prepare_ddp_module(self.state.model, self._find_unused_parameters)
 
+        # The model would need to be torch.compile()'d after being wrapped in a distributed strategy
+        # to take advantage of any graph breaks.
+        if is_torch_2_0 and not is_model_compiled and compile_config is not None:
+            compiled_model = torch.compile(  # pyright: ignore [reportGeneralTypeIssues]
+                self.state.model, **compile_config)
+            self.state.model = compiled_model._orig_mod
+            self.state.model.forward = compiled_model.dynamo_ctx(self.state.model.forward)
+            is_model_compiled = True
+            # update local_hparams to ensure the `is_model_compiled` is set correctly for
+            # debugging purpose and for unit test.
+            if self.auto_log_hparams:
+                self.local_hparams['is_model_compiled'] = is_model_compiled
+        elif not is_torch_2_0 and compile_config is not None:
+            log.warning(f'`torch.compile` is supported for PyTorch 2.0 or higher.' +
+                        f'Either update your PyTorch version or disable parameter by providing ' +
+                        f'`compile_config` to `None`.')
```

Comment:
```suggestion
            warnings.warn(f'`torch.compile` is supported for PyTorch 2.0 or higher.' +
                        f'Either update your PyTorch version or disable parameter by providing ' +
                        f'`compile_config` to `None`.')
```

Reply:
We use warnings.warn when it's a user parameter specification issue, eg there's an actionable step

Reply:
Done in https://github.com/mosaicml/composer/pull/2118/commits/d0e8481b1a3de2bb20239802a9adfac217fb45cc.